src/contracts/core/Slasher.sol

```diff
@@ -27,7 +28,218 @@ import "@openzeppelin-upgrades/contracts/proxy/utils/Initializable.sol";
+    function executeSlashing(address operator, IStrategy[] memory strategies, int256 epoch) external {
+        // TODO: decide if this needs a stonger condition. e.g. the epoch must be further in the past
+        require(epoch < SlashingAccountingUtils.currentEpoch(),
+            "must slash for a previous epoch");
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            uint256 bipsToSlash = pendingSlashedBips[operator][strategy][epoch];
+            pendingSlashedBips[operator][strategy][epoch] = 0;
+            _slashShares(operator, strategy, epoch, bipsToSlash);
+        }
+    }
+
+    function _slashShares(address operator, IStrategy strategy, int256 epoch, uint256 bipsToSlash) internal {
+        int256 lastSlashedEpoch = lastSlashed(operator, strategy);
+        // TODO: again note that we need something like the first epoch being epoch 1 here, to allow actually slashing in the first epoch
+        require(epoch > lastSlashedEpoch, "slashing must occur in strictly ascending epoch order");
+        uint256 scalingFactorBefore = shareScalingFactor(operator, strategy);
+        uint256 scalingFactorAfter = SlashingAccountingUtils.findNewScalingFactor({
+            scalingFactorBefore: scalingFactorBefore,
+            bipsToSlash: bipsToSlash
+        });
+        // update storage to reflect the slashing
+        slashedEpochHistory[operator][strategy].push(epoch);
+        _shareScalingFactor[operator][strategy] = scalingFactorAfter;
+        // TODO: note that this behavior risks off-by-one errors. it needs to be clearly defined precisely how the historical storage is supposed to work
+        shareScalingFactorHistory[operator][strategy][epoch] = scalingFactorAfter;
+        // TODO: events!
+    }
+
+    // Mapping: Operator => Strategy => epoch => AVS => requested slashed bips
+    mapping(address => mapping(IStrategy => mapping(int256 => mapping(address => uint256)))) public requestedSlashedBips;
+
+    // Mapping: Operator => Strategy => epoch => pending slashed amount, where pending slashed amount is the
+    // amount that will be slashed when slashing is executed for the current epoch, assuming no existing requests are cancelled or nullified. summed over all AVSs
+    mapping(address => mapping(IStrategy => mapping(int256 => uint256))) public pendingSlashedBips;
+
+    // @notice fetches the bips slashable by an AVS for the shares of a certain strategy delegated to a certain operator for a certain epoch
+    function getSlashableBips(
+        address operator, 
+        address avs, 
+        IStrategy strategy, 
+        int256 epoch
+    ) public pure returns (uint256) {
+        // TODO: this is a stub. it needs implementation
+        return 10;
+    }
+
+    // @notice fetches the maximum slashing rate -- expressed in bips -- by an AVS for the shares of a certain strategy for a certain epoch
+    function getMaxSlashingRate(
```

Comment:
Noting this will need to be an interface added to middleware ServiceManagerBase as well as `createSlashingRequest`,  `reduceRequestedSlashedBips`, `executeSlashing`